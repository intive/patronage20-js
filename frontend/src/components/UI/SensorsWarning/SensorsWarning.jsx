import React, { useState, useEffect } from 'react'
import { useSelector } from 'react-redux'

import Snackbar from '@material-ui/core/Snackbar'
import { makeStyles } from '@material-ui/core/styles'
import Alert from '@material-ui/lab/Alert'
import IconButton from '@material-ui/core/IconButton'
import Close from '@material-ui/icons/Close'

const useStyles = makeStyles((theme) => ({
  close: {
    padding: theme.spacing(0.5)
  }
}))

export default function SensorsWarning () {
  const classes = useStyles()
  const refreshError = useSelector((state) => state.sensor.refreshError)

  const [open, setOpen] = useState(false)

  useEffect(() => {
    setOpen(!!refreshError)
  }, [refreshError])

  const handleClose = () => {
    setOpen(false)
  }

  return (
    <Snackbar
      anchorOrigin={{ vertical: 'top', horizontal: 'center' }}
      key='top, center'
      open={open}
    >
      <Alert
        variant='filled'
        severity='warning'
        action={
          <IconButton
            aria-label='close'
            color='inherit'
            className={classes.close}
            onClick={handleClose}
          >
            <Close />
          </IconButton>
        }
      >
        Odświeżenie stanu czujników nie powiodło się.
      </Alert>
    </Snackbar>
  )
}
