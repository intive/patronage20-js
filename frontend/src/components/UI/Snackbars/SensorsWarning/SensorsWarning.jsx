import { useState, useEffect } from 'react'
import { useSelector } from 'react-redux'
import { useSnackbar } from 'notistack'

export default function SensorsWarning () {
  const { enqueueSnackbar, closeSnackbar } = useSnackbar()
  const refreshError = useSelector((state) => state.sensor.refreshError)

  const [key, setKey] = useState(null)

  useEffect(() => {
    if (refreshError && !key) {
      setKey(enqueueSnackbar('Odświeżenie stanu czujników nie powiodło się.', {
        variant: 'warning',
        persist: true,
        onExiting: () => setKey(null)
      }))
    } else if (!refreshError && key) {
      closeSnackbar(key)
      setKey(null)
    }
  }, [refreshError])

  return null
}
