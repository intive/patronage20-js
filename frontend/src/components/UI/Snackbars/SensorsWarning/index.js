import SensorsWarning from './SensorsWarning.jsx'

export default SensorsWarning
