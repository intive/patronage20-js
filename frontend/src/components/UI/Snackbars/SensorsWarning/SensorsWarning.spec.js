import React from 'react'
import { create } from 'react-test-renderer'
import { render, screen } from '@testing-library/react'
import SensorsWarning from './index'
import { SnackbarProvider } from 'notistack'

import { Provider } from 'react-redux'

import configureStore from 'redux-mock-store'

jest.mock('notistack', () => ({
  ...jest.requireActual('notistack')
}))

describe('<SensorsWarning />', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  it('renders SensorsWarning component', () => {
    const initialState = { sensor: { } }
    const mockStore = configureStore()
    const store = mockStore(initialState)

    const root = create(
      <Provider store={store}>
        <SnackbarProvider>
          <SensorsWarning />
        </SnackbarProvider>
      </Provider>
    )
    expect(root.toJSON()).toMatchSnapshot()
  })

  it('should show a snackbar if refreshError is not null', () => {
    const initialState = { sensor: { refreshError: 'error' } }
    const mockStore = configureStore()
    const store = mockStore(initialState)

    render(
      <Provider store={store}>
        <SnackbarProvider>
          <SensorsWarning />
        </SnackbarProvider>
      </Provider>
    )

    expect(screen.getByText('Odświeżenie stanu czujników nie powiodło się.')).not.toEqual(null)
  })

  it('should not show a new snackbar if one is already open', () => {
    const notistack = require('notistack')

    const enqueueSnackbar = jest.fn()
    jest.spyOn(notistack, 'useSnackbar').mockImplementation(() => { return { enqueueSnackbar } })

    const setKey = jest.fn()
    const useStateMock = () => ['testKey', setKey]
    jest.spyOn(React, 'useState').mockImplementation(useStateMock)

    const initialState = { sensor: { refreshError: 'error' } }
    const mockStore = configureStore()
    const store = mockStore(initialState)

    render(
      <Provider store={store}>
        <SnackbarProvider>
          <SensorsWarning />
        </SnackbarProvider>
      </Provider>
    )

    expect(enqueueSnackbar).toHaveBeenCalledTimes(0)
    expect(setKey).toHaveBeenCalledTimes(0)
  })

  it('should close a snackbar if snackbar is open and refreshError is set to null', () => {
    const notistack = require('notistack')

    const closeSnackbar = jest.fn()
    jest.spyOn(notistack, 'useSnackbar').mockImplementation(() => { return { closeSnackbar } })

    const setKey = jest.fn()
    const useStateMock = () => ['testKey', setKey]
    jest.spyOn(React, 'useState').mockImplementation(useStateMock)

    const initialState = { sensor: { refreshError: null } }
    const mockStore = configureStore()
    const store = mockStore(initialState)

    render(
      <Provider store={store}>
        <SnackbarProvider>
          <SensorsWarning />
        </SnackbarProvider>
      </Provider>
    )

    expect(closeSnackbar).toHaveBeenCalledTimes(1)
    expect(setKey).toHaveBeenCalledTimes(1)
  })
})
