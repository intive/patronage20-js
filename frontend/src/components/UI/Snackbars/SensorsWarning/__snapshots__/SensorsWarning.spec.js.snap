// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SensorsWarning /> renders SensorsWarning component 1`] = `null`;
